packages/caver-wallet/src/keyring/keyring.js

```diff
@@ -0,0 +1,869 @@
+     */
+    /**
+     * options can include below
+     * {
+     *   salt: ...,
+     *   iv: ...,
+     *   kdf: ...,
+     *   dklen: ...,
+     *   c: ...,
+     *   n: ...,
+     *   r: ...,
+     *   p: ...,
+     *   cipher: ...,
+     *   uuid: ...,
+     *   cipher: ...,
+     * }
+     */
+    encrypt(password, options = {}) {
+        let keyring = []
+        let isRoleBased = false
+
+        for (let i = KEY_ROLE.ROLE_TRANSACTION_KEY; i < KEY_ROLE.ROLE_LAST; i++) {
+            const roledKey = this._key[i]
+            if (i > KEY_ROLE.ROLE_TRANSACTION_KEY && roledKey.length > 0) isRoleBased = true
+            keyring.push(encryptKey(roledKey, password, options))
+        }
+
+        if (!isRoleBased) keyring = keyring[0]
+
+        return formatEncrypted(4, this._address, keyring, options)
+    }
+
+    /**
+     * encrypts a keyring and returns a keystore v3 object.
+     *
+     * @param {string} password The password to be used for keyring encryption. The encrypted key store can be decrypted with this password.
+     * @param {object} options The options to use when encrypt a keyring. See `keyring.encrypt` for more detail about options.
+     * @return {object}
+     */
+    encryptV3(password, options) {
+        const notAvailableError = `This keyring cannot be encrypted keystore v3. use 'keyring.encrypt(password)'.`
+        if (this._key[KEY_ROLE.ROLE_TRANSACTION_KEY].length > 1) throw new Error(notAvailableError)
+
+        for (let i = KEY_ROLE.ROLE_ACCOUNT_UPDATE_KEY; i < KEY_ROLE.ROLE_LAST; i++) {
+            if (this._key[i].length > 0) {
+                throw new Error(notAvailableError)
+            }
+        }
+
+        options = options || {}
+
+        const crypto = encryptKey(this._key[0][0], password, options)
+
+        return formatEncrypted(3, this._address, crypto, options)
+    }
+
+    _validateOptionsForUpdate(options = []) {
+        // { threshold: 1, weight: [1, 1] } => [{ threshold: 1, weight: [1, 1] }]
+        if (!_.isArray(options)) options = [options]
+
+        for (let i = 0; i < this._key.length; i++) {
+            // Validation for options obejct will be operated in AccountKeyWeightedMultiSig class.
+            if (options[i] && Object.keys(options[i]).length > 0) continue
+
+            let optionToAdd
+            if (this._key[i].length > 1) {
+                // default option when option is not set
+                optionToAdd = { threshold: 1, weight: Array(this._key[i].length).fill(1) }
+            } else {
+                // AccountKeyPublic does not need option
+                optionToAdd = {}
+            }
+
+            if (options[i]) {
+                options[i] = optionToAdd
+            } else {
+                options.push(optionToAdd)
+            }
+        }
+        return options
+    }
+}
+
+Keyring.PrivateKey = PrivateKey
+Keyring.role = KEY_ROLE
+
+module.exports = Keyring
+
+/**
+ * Formatting key paramter for member variable in Keyring.
```

Comment:
It seems unclear. This function does not belong to Keyring. `member variable` is also somewhat misleading.

- Why is this function necessary?
- What is the purpose of this function?

Reply:
I modified description. :)